libpod/rootless_cni_linux.go

```diff
@@ -255,46 +269,65 @@ func startRootlessCNIInfraContainer(ctx context.Context, r *Runtime) (*Container
 		return nil, err
 	}
 	g.RemoveMount("/proc")
-	procMount := spec.Mount{
-		Destination: "/proc",
-		Type:        "bind",
-		Source:      "/proc",
-		Options:     []string{"rbind", "nosuid", "noexec", "nodev"},
-	}
-	g.AddMount(procMount)
-	// Mount CNI networks
-	etcCNINetD := spec.Mount{
-		Destination: "/etc/cni/net.d",
-		Type:        "bind",
-		Source:      r.config.Network.NetworkConfigDir,
-		Options:     []string{"ro", "bind"},
-	}
-	g.AddMount(etcCNINetD)
 
-	inspectData, err := newImage.Inspect(ctx)
-	if err != nil {
+	// need writable /run
+	run := spec.Mount{
+		Destination: "/run",
+		Type:        "tmpfs",
+		Source:      "none",
+		Options:     []string{"rw", "nosuid", "nodev"},
+	}
+	g.AddMount(run)
+
+	// mount var as tmpfs
+	// cni tries to write to /var/lib/cni however we cannot mount there since
+	// it might not exists and we have no permission to create this directory
+	cni := spec.Mount{
+		Destination: "/var",
+		Type:        "tmpfs",
+		Source:      "none",
+		Options:     []string{"rw", "nosuid", "nodev"},
+	}
+	g.AddMount(cni)
+
+	g.SetProcessArgs([]string{rootlesscni.RootlessCNIInfraCmd, "sleep"})
+
+	rootfs := filepath.Join(r.GetStore().RunRoot(), rootlessCNIInfraContainerName)
+	if err := os.MkdirAll(rootfs, 0700); err != nil {
 		return nil, err
 	}
-	imageEnv, err := env.ParseSlice(inspectData.Config.Env)
+
+	// get the current path this executable so we can mount it
+	podmanexe, err := os.Executable()
 	if err != nil {
 		return nil, err
 	}
-	for k, v := range imageEnv {
-		g.AddProcessEnv(k, v)
+	podman := spec.Mount{
+		// mount with different name to trigger the reexec for rooless-cni-infra
+		Destination: rootlesscni.RootlessCNIInfraCmd,
+		Type:        "bind",
+		Source:      podmanexe,
+		Options:     []string{"ro", "bind"},
 	}
-	if len(inspectData.Config.Cmd) == 0 {
-		return nil, errors.Errorf("rootless CNI infra image %q has no command specified", imageName)
+	g.AddMount(podman)
+
+	// bind mount the rootfs in the userns read only
+	if err := mount.Mount("/", rootfs, "bind", "rbind,rprivate,ro"); err != nil {
+		return nil, errors.Wrapf(err, "failed to mount rootfs for %s", rootlessCNIInfraContainerName)
 	}
-	g.SetProcessArgs(inspectData.Config.Cmd)
 
-	var options []CtrCreateOption
-	options = append(options, WithRootFSFromImage(newImage.ID(), imageName, imageName))
-	options = append(options, WithCtrNamespace(rootlessCNIInfraContainerNamespace))
-	options = append(options, WithName(rootlessCNIInfraContainerName))
-	options = append(options, WithPrivileged(true))
-	options = append(options, WithSecLabels([]string{"disable"}))
-	options = append(options, WithRestartPolicy("always"))
-	options = append(options, WithNetNS(nil, false, "slirp4netns", nil))
+	g.SetRootReadonly(true)
+
+	options := []CtrCreateOption{
+		WithRootFS(rootfs),
+		WithCtrNamespace(rootlessCNIInfraContainerNamespace),
+		WithName(rootlessCNIInfraContainerName),
+		WithPrivileged(true),
+		WithSecLabels([]string{"disable"}),
```

Comment:
selinux should be disabled for this container so I am not sure why selinux is blocking the access to this file

Reply:
The problem is the iptables command is confined. The question is who launched it? And why did it transition.